_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,1079 @@
+---
+title: "How to debug an ARM Cortex-M HardFault"
+description: "A step by step guide of how to investigate what led to a fault on a Cortex-M device with examples and strategies about how to recover"
+tag: [cortex-m]
+author: chris
+---
+
+At one point or another when working with an ARM Cortex-M MCU, you will hit something like:
+
+```shell
+(gdb) backtrace
+#0  HardFault_Handler() at startup.c:136
+#1  <signal handler called>
+```
+
+<!-- excerpt start -->
+
+In this article we will demystify how to determine what went wrong when an ARM Cortex-M fault occurs. We will walk step by step through what the registers to inspect and what the values actually mean. We will discuss debug tricks that can be used for some of the most challenging classes of HardFaults, how how to automate the fault analysis, and how to recover from a fault without rebooting the system. Finally, we will apply what we have learned in the article and walk through a couple practical examples.
```

Comment:
Couple typos here, and more verbose than it needs to be:

> In this article, we explain how to debug faults on ARM Cortex-M based devices. In the process, we learn about fault registers, automate fault analysis, and figure our ways to recover from some faults without rebooting. We include practical examples, with a step by step walkthrough on how to investigate them.

Reply:
sounds great, thanks!